fili-sql/src/main/java/com/yahoo/bard/webservice/sql/DruidQueryToSqlConverter.java

```diff
@@ -402,40 +415,99 @@ protected RelNode convertDruidQueryToRelNode(
+                            NO_OFFSET,
+                            getLimit(druidQuery),
+                            getSort(builder, druidQuery, apiToFieldMapper, sqlTable.getTimestampColumn())
+            )
+                    .build();
         }
 
-        RelBuilder builder = calciteHelper.getNewRelBuilder(sqlTable.getSchemaName(), sqlTable.getCatalog());
-        builder = builder.scan(sqlTable.getName());
-        return builder
-                .filter(
-                        getAllWhereFilters(builder, druidQuery, apiToFieldMapper, sqlTable.getTimestampColumn())
-                )
-                .aggregate(
-                        builder.groupKey(getAllGroupByColumns(
-                                builder,
-                                druidQuery,
-                                apiToFieldMapper,
-                                sqlTable.getTimestampColumn()
-                        )),
-                        getAllQueryAggregations(builder, druidQuery, apiToFieldMapper)
-                )
-                .project(
-                        (Iterable) ImmutableList.builder()
-                                .addAll(builder.fields())
-                                .addAll(getPostAggregations(builder, druidQuery, apiToFieldMapper))
-                                .build()
-                )
-                .filter(
-                        getHavingFilter(builder, druidQuery, apiToFieldMapper)
-                )
-                .sortLimit(
-                        NO_OFFSET,
-                        getLimit(druidQuery),
-                        getSort(builder, druidQuery, apiToFieldMapper, sqlTable.getTimestampColumn())
-                )
-                .build();
+        builder.push(queryRelNode);
+        if (druidQuery instanceof TopNQuery) {
+            long threshold = ((TopNQuery) druidQuery).getThreshold();
+            TopNMetric topNMetric = ((TopNQuery) druidQuery).getMetric();
+            Set<SqlKind> directions = new HashSet<>();
+            if (topNMetric.getType().equals(TopNMetric.TopNMetricType.INVERTED)) {
+                directions.add(SqlKind.NULLS_FIRST);
+            } else {
+                directions.add(SqlKind.DESCENDING);
+                directions.add(SqlKind.NULLS_LAST);
+            }
+
+            Object metric = topNMetric.getMetric();
+            while (metric instanceof TopNMetric) {
+                metric = ((TopNMetric) metric).getMetric();
+            }
+            if (metric == null) {
+                throw new IllegalStateException("null metric found for the topN query");
+            }
+            String metricName = (String) metric;
+            int timePartFunctions = sqlTimeConverter.timeGrainToDatePartFunctions(druidQuery.getGranularity()).size();
+            int groupBys = druidQuery.getDimensions().size() + timePartFunctions;
+            List<RexNode> timePartitions = builder.fields().subList(druidQuery.getDimensions().size(), groupBys);
```

Comment:
I think this would be cleaner if it was like 
```suggestion
            int dimensions = druidQuery.getDimensions().size();
            List<RexNode> timePartitions = builder.fields().subList(dimensions, dimension + timePartFunctions);
```

Reply:
👍